dm/syncer/syncer.go

```diff
@@ -2042,7 +2044,7 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
 	param := &genDMLParam{
 		targetTableID:   utils.GenTableID(targetTable),
 		data:            prunedRows,
-		originalData:    rows,
+		originalData:    extRows,
```

Comment:
maybe we should keep using rows as `originalData`, and apply `extRows` to `prunedColumns`

Reply:
The `update/delete` events will check that the lengths of `originalData` and `sourceTableInfo.Columns` are equal.
https://github.com/yufan022/ticdc/blob/dm_2111/dm/syncer/dml.go#L209

Since `sourceTableInfo.Columns` from the downstream table for first time, that will contains `table/source/schema` columns info.

Reply:
I still perfer to not change the originData because we should keep raw data from binlog not affected by `ExtendColumn`. so the logic of the check [here](https://github.com/yufan022/ticdc/blob/152ae257fd8fabde8caeaba47ef71e250b96a12d/dm/syncer/dml.go#L209) could be updated.

because https://github.com/pingcap/ticdc/pull/3168 is merged , i think we shoud check the error: `ErrSyncerUnitDMLColumnNotMatch` by campare input data `param.data` with `downstreamTableInfo.Columns` ?
ptal @lance6716 @GMHDBJD

Reply:
I agree with you and there has another way: If `extCols > 0` then check `len(originData) != len(extCols)`.

Reply:
i can't understand why need to check `len(originData) != len(extCols)`, seems they are always different when there are extCols 😂, my point is just to add extrow to `prunedColumns` so that we can make sure `extrow` + `prunedColumns` have same column count with downstream

Reply:
sorry typo fault. check `if extCols > 0` then check `(len(originData) + len(extCols)) != len(ti.Columns) {`

Reply:
For update/delete scene, we need change `delete from a where id = 1` to `delete from a where id = 1 and table = x and source = x and schema = x`.

`where id = 1` comes from `originalData`, So if we can't modify `originData`, the situation will be a little complicated.

Reply:
> delete/update from source_table where id = 1 to delete from target_table where id = 1 and table = x and source = x and schema = x

LGTM

> if we can't append extCols to originalData, the situation will be a little complicated.

seems we only need do some changes in func `whereColumnsAndValues` just add extCols and extValues

Reply:
seems just need add extValues, because [sourceTableInfo](https://github.com/yufan022/ticdc/blob/2b46cf74d8dc8e32866170188241ab15555ba605/dm/syncer/syncer.go?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L2051) is from downStream, it always contains extendColumn info.

I think if the downstream table only has more extended columns than the upstream table, we don't need use `dmctl operate-schema set` to set table schema, just use the downstream [table](https://github.com/yufan022/ticdc/blob/e4da0cfc5d469b447f86abd2e3f94c65cd865919/dm/syncer/syncer.go#L715) directly. Is that Is correct?

Reply:
> seems just need add extValues

i don't think so, because the extend col may not  be an index column in downstream, and if it is, the genereated sql will  not  contain the extend cols in current [impl](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L631)

> I think if the downstream table only has more extended columns than the upstream table, we don't need use dmctl operate-schema set to set table schema, just use the downstream table directly. Is that correct?

yes, we now always use downstream table strucure to `gensql`

Reply:
> i don't think so, because the extend col may not be an index column in downstream, and if it is, the genereated sql will not contain the extend cols in current impl

usually downstream table will change the primary key to union primary key like:

source: `PRIMARY KEY (id)`
target: `PRIMARY KEY(id,source_table)`

In this scene, that [indexColumns.Columns](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L381) has `id, source_table`, [columns](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L382) has extend column, but [data](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L383) does not contains extend columns, just need add extVals to data.

source: `PRIMARY KEY (id)`
target: `does not contains extend column index or primary key`

In this scene, [columns](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L635) already contains extend column name, just need add extVals to [values](https://github.com/pingcap/ticdc/blob/8814deda74767c7094294441c9c6517244dd824e/dm/syncer/dml.go#L647).

In the above two scenarios, it seems that we only need to add extVals.

Reply:
i agree with lance6716's comments here, i will resolve this conversation as the comments below have summed it up prety well